client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,251 @@
+		let mut span_data = self.spans.lock();
+		if let Some(span) = span_data.get_mut(&id) {
+			span.entered.push(Instant::now() - self.timestamp);
+		}
+	}
+
+	fn exit(&self, span: &Id) {
+		if let Some(s) = self.spans.lock().get_mut(span) {
+			self.current_span.exit();
+			s.exited.push(Instant::now() - self.timestamp)
+		}
+	}
+}
+
+/// Holds a reference to the client in order to execute the given block
+/// and record traces for the supplied targets (eg. "pallet,frame,state")
+pub struct BlockExecutor<Block: BlockT, Client> {
+	client: Arc<Client>,
+	block: Block::Hash,
+	targets: Option<String>,
+}
+
+impl<Block, Client> BlockExecutor<Block, Client>
+	where
+		Block: BlockT + 'static,
+		Client: HeaderBackend<Block> + BlockBackend<Block> + ProvideRuntimeApi<Block>
+		+ Send + Sync + 'static,
+		Client::Api: Metadata<Block, Error=sp_blockchain::Error>,
+{
+	/// Create a new `BlockExecutor`
+	pub fn new(client: Arc<Client>, block: Block::Hash, targets: Option<String>) -> Self {
+		Self { client, block, targets }
+	}
+
+	/// Execute block, recording all spans and events belonging to `Self::targets`
+	pub fn trace_block(&self) -> Result<BlockTrace, String> {
+		// Prepare block
+		let id = BlockId::Hash(self.block);
+		let extrinsics = self.client.block_body(&id)
+			.map_err(|e| format!("Invalid block id: {:?}", e))?
+			.ok_or("Block not found".to_string())?;
+		let mut header = self.client.header(id)
+			.map_err(|e| format!("Invalid block id: {:?}", e))?
+			.ok_or("Header not found".to_string())?;
+		let parent_hash = header.parent_hash().clone();
+		let parent_id = BlockId::Hash(parent_hash.clone());
+		// Pop digest else RuntimePanic due to: 'Number of digest items must match that calculated.'
+		header.digest_mut().pop();
+		let block = Block::new(header, extrinsics);
+
+		let targets = if let Some(t) = &self.targets { t } else { DEFAULT_TARGETS };
+		let spans = Arc::new(Mutex::new(HashMap::new()));
+		let events = Arc::new(Mutex::new(Vec::new()));
+		let sub = BlockSubscriber::new(targets, spans.clone(), events.clone());
+		let dispatch = Dispatch::new(sub);
+
+		if let Err(e) = dispatcher::with_default(&dispatch, || {
+			let span = tracing::info_span!(
+				target: TRACE_TARGET,
+				"trace_block",
+			);
+			let _enter = span.enter();
+			self.client.runtime_api().execute_block(&parent_id, block)
+		}) {
+			return Err(format!("Error executing block: {:?}", e));
+		}
+		drop(dispatch);
+		let spans = Arc::try_unwrap(spans)
+			.map_err(|_| "Unable to unwrap spans".to_string())?;
+		let events = Arc::try_unwrap(events)
+			.map_err(|_| "Unable to unwrap spans".to_string())?;
+		let spans: Vec<Span> = spans.into_inner().into_iter().map(|(_, s)| s.into()).collect();
+		// First filter out any spans that never entered
+		let spans: Vec<Span> = spans.into_iter().filter(|s| !s.entered.is_empty()).collect();
+		// Patch wasm identifiers
+		let mut spans: Vec<Span> = spans.into_iter().filter_map(|s| patch_and_filter(s, targets)).collect();
```

Comment:
```suggestion
		let spans: Vec<Span> = spans.lock().drain().map(|(_, s)| s.into()).into_iter().filter(|s| !s.entered.is_empty()).filter_map(|s| patch_and_filter(s, targets)).collect();
```

You should get it ;) This way we will not have to do 3 collects, only one.